tuf/tuf.go

```diff
@@ -632,47 +598,118 @@ func (tr *Repo) VerifyCanSign(roleName string) error {
 	return signed.ErrNoKeys{KeyIDs: role.ListKeyIDs()}
 }
 
+// used for walking the targets/delegations tree, potentially modifying the underlying SignedTargets for the repo
+type walkVisitorFunc func(*data.SignedTargets, data.DelegationRole) error
+
+// WalkTargets will apply the specified visitor function to iteratively walk the targets/delegation metadata tree,
+// until receiving a ErrStopWalk.  The walk starts from the base "targets" role, and searches for the correct targetPath and/or rolePath
+// to call the visitor function on.
+func (tr *Repo) WalkTargets(targetPath, rolePath string, visitTarget walkVisitorFunc) error {
+	// Start with the base targets role, which implicitly has the "" targets path
+	targetsRole, err := tr.GetBaseRole(data.CanonicalTargetsRole)
+	if err != nil {
+		return err
+	}
+	// Make the targets role have the empty path, when we treat it as a delegation role
+	roles := []data.DelegationRole{
+		{
+			BaseRole: targetsRole,
+			Paths:    []string{""},
+		},
+	}
+
+	for len(roles) > 0 {
+		role := roles[0]
+		roles = roles[1:]
+
+		// Check the role metadata
+		signedTgt, ok := tr.Targets[role.Name]
+		if !ok {
+			// The role meta doesn't exist in the repo so continue onward
+			continue
+		}
+
+		// We're at a prefix of the desired role subtree, so add its delegation role children and continue walking
+		if strings.HasPrefix(rolePath, role.Name+"/") {
+			roles = append(roles, signedTgt.GetValidDelegations(role)...)
+			continue
+		}
+
+		// Determine whether to visit this role or not:
+		// If the paths validate against the specified targetPath and the rolePath is empty or is in the subtree
+		if isValidPath(targetPath, role) && isAncestorRole(role.Name, rolePath) {
+			// If we had matching path or role name, visit this target and determine whether or not to keep walking
+			err := visitTarget(signedTgt, role)
+			switch err.(type) {
+			case ErrStopWalk:
+				// If the visitor function signalled a stop, return nil to finish the walk
+				return nil
+			case ErrContinueWalk:
+				// If the visitor function signalled to continue, add this role's delegation to the walk
+				roles = append(roles, signedTgt.GetValidDelegations(role)...)
+			default:
+				// Return out if we got a different error or nil
+				return err
+			}
+
+		}
+	}
+	return nil
+}
+
+// helper function that returns whether the candidateChild role name is an ancestor or equal to the candidateAncestor role name
+// Will return true if given an empty candidateAncestor role name
+func isAncestorRole(candidateChild, candidateAncestor string) bool {
+	return candidateAncestor == "" || candidateAncestor == candidateChild || strings.HasPrefix(candidateChild, candidateAncestor+"/")
```

Comment:
Should the arguments to the `strings.HasPrefix` part of this be reversed?  Since if we are at role `targets` with `rolePath = "targets/releases"`, we want to check if `targets/releases` has the prefix `targets/`.  (Although we seem to never reach this point, since [we already check for that condition and skip, if so, above](https://github.com/docker/notary/pull/577/files#diff-b7b7bae92e6be55133dab8ee3256e014R633)